syntax = "proto2";

message BankAccountProto {
  repeated uint64 p = 1;
  repeated uint64 n = 2;
}

// Deposit.
message BankAccountDepositRequest {
  required int64 amount = 1;
}

message BankAccountDepositReply {
}

// Withdraw.
message BankAccountWithdrawRequest {
  required int64 amount = 1;
}

message BankAccountWithdrawReply {
}

// Get.
message BankAccountGetRequest {
}

message BankAccountGetReply {
  required int64 value = 1;
}

// Request and reply.
message BankAccountTxnRequest {
  enum Type {
    DEPOSIT = 0;
    WITHDRAW = 1;
    GET = 2;
  }
  required Type type = 1;
  optional BankAccountDepositRequest deposit = 2;
  optional BankAccountWithdrawRequest withdraw = 3;
  optional BankAccountGetRequest get = 4;
}

message BankAccountTxnReply {
  enum Result {
    COMMITTED = 0;
    ABORTED = 1;
  }
  required Result result = 1;
  optional BankAccountDepositReply deposit = 2;
  optional BankAccountWithdrawReply withdraw = 3;
  optional BankAccountGetReply get = 4;
}
